system/jlib/jsocket.cpp

```diff
@@ -466,6 +467,200 @@ extern jlib_decl bool queryKeepAlive(int &time, int &intvl, int &probes)
+                CriticalBlock block(queryDNSCS);
+                if (!gaPtrList.empty())
+                {
+                    std::list<GetAddrInfoThread *>::iterator iter = gaPtrList.begin();
+                    std::list<GetAddrInfoThread *>::iterator end  = gaPtrList.end();
+                    while (iter != end)
+                    {
+                        GetAddrInfoThread *pItem = *iter;
+                        if ( (pItem->thrdHasEnded()) && (pItem->join(20)) )
+                        {
+                            // TODO: should we log failures ? (but not here inside CS)
+                            delete pItem;
+                            iter = gaPtrList.erase(iter);
+                        }
+                        else
+                            ++iter;
+                    }
+                }
+            }
+            if (sem.wait(10))
+                break;
+        }
+        return 0;
+    }
+
+};
+
+static std::atomic<UseUDE> disableDNSTimeout { UNINIT };
+
+static Owned<AddrInfoReaperThread> addrinforeaperthrd;
+
+static bool queryDNSTimeout()
+{
+    UseUDE state = disableDNSTimeout.load();
+    if (state == UNINIT)
+    {
+        CriticalBlock block(queryDNSCS);
+        state = disableDNSTimeout.load();
+        if (state == UNINIT)
+        {
+#ifdef _CONTAINERIZED
+            Owned<IPropertyTree> expert;
+#else
+            Owned<IPropertyTree> envtree;
+            IPropertyTree *expert = nullptr;
+#endif
+            try
+            {
+#ifdef _CONTAINERIZED
+                expert.setown(getGlobalConfigSP()->getPropTree("expert"));
+#else
+                // MCK - without this many components will not have a global prop
+                envtree.setown(getHPCCEnvironment());
+                if (envtree)
+                    expert = envtree->queryPropTree("Software/Globals");
+#endif
+            }
+            catch (IException *e)
+            {
+                e->Release();
+            }
+            catch (...)
+            {
+            }
+            state = ENABLED;
+            if (expert)
+            {
+                bool isDisabled = expert->getPropBool("@disableDNSTimeout", false);
+                if (isDisabled)
+                    state = DISABLED;
+            }
+            disableDNSTimeout = state;
+
+            if (state == ENABLED)
+            {
+                if (!addrinforeaperthrd)
+                {
+                    addrinforeaperthrd.setown(new AddrInfoReaperThread());
+                    addrinforeaperthrd->start(false);
+                }
+            }
+        }
+    }
+
+    if (state == ENABLED)
+        return true;
+
+    return false;
+}
+
+MODULE_INIT(INIT_PRIORITY_STANDARD)
+{
+    return true;
+}
+
+MODULE_EXIT()
+{
+    // BUG: this (and all other MODULE_EXITs) are not called for Thor and Roxie because they are
+    //      stopped via SIGTERM signal and thus exit() and the atexit handlers are not called ...
```

Comment:
Is this as designed/intended ?  I want to make sure a reaper thread is terminated before the process ends, but it seems MODULE_EXIT() is not called when Thor and Roxie end.
If I call exit() on SIGTERM then its called, but some other issues occur at process end that need to be looked into.